Add repair step to load missing apps before upgrade

VicDeo opened the pull request:
## Description
Before update repair step

## Related Issue
owncloud/market#39
https://github.com/owncloud/core/issues/26227

## Motivation and Context
Allow seamless app unbundling  for the next releases

## How Has This Been Tested?
Testing in progress

## Screenshots (if appropriate):

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

VicDeo committed:
Add repair step to load missing apps before upgrade

PVince81 commented on lib/private/Repair/Apps.php:
```diff
@@ -0,0 +1,100 @@
+			if ($isMarketEnabled){
+				$this->loadApp('market');
+				foreach ($missingApps as $app) {
+					$output->info("Repairing missing app: $app");
```
I wouldn't call it "repairing" because it sounds like it's broken.

Migrating from 9.1 to 10.0 it is an expected step to migrate the app.
Maybe call this "Fetching app from market"

PVince81 commented on lib/private/Repair/Apps.php:
```diff
@@ -0,0 +1,100 @@
+						IRepairStep::class . '::repairAppStoreApps',
+						new GenericEvent($app)
+					);
+					$this->appManager->enableApp($app);
```
Not sure if this is safe, I'd suggest to first fetch all apps and only enable them all at the end ?

Just in case one app in the middle breaks everything...

PVince81 commented:
Code makes sense so far

PVince81 commented on lib/private/Repair/Apps.php:
```diff
@@ -0,0 +1,100 @@
+				foreach ($missingApps as $app) {
+					$output->info("Repairing missing app: $app");
+					$this->eventDispatcher->dispatch(
+						IRepairStep::class . '::repairAppStoreApps',
```
Hmmm, what if it fails ? Does it abort correctly ?

Also need to display the error to the admin so they can find if it's a connectivity issue or maybe they're unlucky and the app doesn't exist on the market.

In case the app doesn't exist on the market then it's the situation from https://github.com/owncloud/core/pull/27047 where someone deleted the code of a third party app and is trying to upgrade.

VicDeo committed:
App restoring improvements

VicDeo commented:
@PVince81  Regarding your question "do you pull the app from the market before the DB update ?"
Works for clean install too:
1. install 10rc1
2. Enable user_ldap
3. Replace with daily
4. Apply core/market patches
5. Kill user_ldap
6. Update
7. App is alive again

Detecting an exact error requires additional exception subclassing in the market app

PVince81 commented on lib/private/Repair/Apps.php:
```diff
@@ -0,0 +1,111 @@
+					IRepairStep::class . '::repairAppStoreApps',
+					new GenericEvent($app)
+				);
+				$this->appManager->enableApp($app);
```
I'm not sure whether it's the right time to enable the app here. From what I remember the update process has a specific moment where it enables all apps, maybe as part of upgrading apps, not sure. Can you verify that this fits in the workflow ? Ideal is to have the exact same workflow as before and not enable apps too early or too late which has a risk of influencing / modifying update behavior.

VicDeo replied:
@PVince81 Update tries to enable only apps that were disabled by update (3rd party or incompatible)
I haven't managed to find the reason but the app code is restored but the app is disabled otherwise.

VicDeo committed:
Fetch all apps

VicDeo committed:
Handle no connection to market

VicDeo commented:
- [x] Check for apps that have missing app code, store in a list
- [x] Check for apps that are present but are incompatible, store in same list
- [x] Check if the market app is enabled
- [x] If it is enabled - iterate though the list of enabled apps, updating the apps one by one. Store failures in a separate list.
- [ ] but don't enable - still enabling... 
if there is no connection to market: fallback to the disabled case
- [x] If it is disabled: put all missing/incompatible apps into the list of failures at once
- [x]  if there are failures in a list: 
fail with message "please install app manually with tarball, or disable app with occ app:disable"
- [ ] When all went well, enable all apps (or only the "authenitcation" ones) <= to be discussed, maybe the update process takes care of that already

- [ ] Handling errors from the market app. 
- [ ] Missing apps need `install` hook while incompatible apps need `update` hook.
- [ ] https://github.com/owncloud/core/pull/27648 is not needed anymore -> revert 
https://github.com/owncloud/market/pull/22 needs to be hooked onto `IRepairStep::class` instead of `Updater::class`

VicDeo committed:
Use different events for missing/present apps. Reenable missing apps later.

PVince81 commented:
Tests:

#### Marketplace unavailable / offline
- [ ] TEST: upgrade with no changes to do: skipped, success
- [ ] TEST: upgrade with outdated app: abort
- [ ] TEST: upgrade with missing app code: abort
- [ ] TEST: admin manually updates local app code: success
- [ ] TEST: admin disables broken app: success

#### Marketplace available but app does not exist there
- [ ] TEST: upgrade with no changes to do: skipped, success
- [ ] TEST: upgrade with outdated app: abort
- [ ] TEST: upgrade with missing app code: abort
- [ ] TEST: admin manually updates local app code: success
- [ ] TEST: admin disables broken app: success

#### Marketplace available, app exists but no new version
- [ ] TEST: upgrade with no changes to do: skipped, success
- [ ] TEST: upgrade with outdated app: abort
- [ ] TEST: upgrade with missing app code: abort
- [ ] TEST: admin manually updates local app code: success
- [ ] TEST: admin disables broken app: success

#### Marketplace available, app exists and new version available
- [ ] TEST: upgrade with no changes to do: skipped, success
- [ ] TEST: upgrade with outdated app: app updated, success
- [ ] TEST: upgrade with missing app code: app updated, success
- [ ] TEST: admin manually updates local app code: success
- [ ] TEST: admin disables broken app: success

PVince81 commented:
- [ ] add specific exception for marketplace connection not available ?

PVince81 commented on lib/private/Repair/Apps.php:
```diff
@@ -0,0 +1,221 @@
+
+			throw new RepairException('Upgrade is not possible');
+		} elseif ($hasNotUpdatedCompatibleApps) {
+			// warn?
```
I'd say yes, warn. Then the admin should be able to upgrade these apps inside the market app later on (or update these manually in case of offline mode).

PVince81 commented:
- [ ] log info message in console when processing apps (especially if market is slow, so admin knows what the update process is currently doing)

PVince81 commented:
- [ ] BUG: update with outdated app that has no new version fails

#### My steps:
1. `occ market:install contacts`
1. `occ app:enable contacts`
1. Edit contacts app and make it incompatible by changing max-version
1. Increase version in version.php
1. `occ app upgrade`

#### Expected result:
Update aborts.

#### Actual result:
Update goes through.

@VicDeo not sure if my testing is wrong.
The market does not provide a new version since I left the app version unchanged.

PVince81 commented on lib/private/Repair/Apps.php:
```diff
@@ -0,0 +1,221 @@
+				);
+
+				foreach ($appsToEnable as $app){
+					$this->appManager->enableApp($app);
```
I suggest we don't enable anything here. Why ? Because these apps were already enabled before, else we wouldn't do any updating of them.

I think this loop is anyway a no-op, it will see the app is already enabled and skip. So better remove it.

VicDeo replied:
@PVince81 in case of missing code app is going into a disabled state after the code is restored. I haven't found the reason, this is a fixup.

PVince81 replied:
Can you add a code comment then to prevent future devs to wonder about the same question ?

PVince81 commented:
Looks good so far apart from the one bug.

I'll continue testing with the planned user_ldap case:

#### user_ldap 10.0 case
Setup LDAP in 9.1 then upgrade to 10.0:

- [ ] TEST: with user_ldap app missing: abort
- [ ] TEST: admin installs beta user_ldap manually: upgrade goes through and finds users
- [ ] TEST: admin disables app (not recommended here): upgrade goes through and finds users

butonic commented on lib/private/Repair/Apps.php:
```diff
@@ -0,0 +1,221 @@
+					'reinstallAppStoreApp'
+				);
+				$hasNotUpdatedCompatibleApps = count($failedCompatibleApps);
+
```
Hm, so the current implementation will try to enable apps that have been updated. https://github.com/owncloud/core/blob/2d000d9eba3815972241515d691ae7b8d49ce0c5/lib/private/App/AppManager.php#L218 currently might throw an \Exception, casing a full stop of this process, and then the rest of this function will not be executed.
1. I think it should throw an AppNotInstalledException instead.
2. That still allows the foreach to be escaped if the exception is thrown. I'd rather move 98-105 out of the try catch. If we try to enable an app that wasn't installed in the first place something has gone wrong and I think it is ok to fail early in that case.

PVince81 replied:
For clarification: if an app's code is missing but the DB contains an entry about an enabled app, do we call this "app is installed but missing code" in the logic ? @VicDeo 

From my understanding this code only runs for apps that either are pending updates or where the code has disappeared (so it was installed at some point, like our future "user_ldap" case).

VicDeo replied:
@PVince81 this PR works with `enabled` apps only
there are three kinds of them:
- enabled, but no code
- enabled, but incompatible
- enabled and compatible

VicDeo replied:
these lines  (reenabling the apps) can be safely removed once https://github.com/owncloud/core/pull/27730 is merged.

PVince81 commented:
I did all tests with `occ upgrade`. Will check how the web UI behaves.
For now showing "sorry, need to disable app" is enough.

butonic commented on lib/private/Repair/Apps.php:
```diff
@@ -0,0 +1,221 @@
+		}
+
+		// Do not put this in else
+		if (!$isMarketEnabled) {
```
The current code will overwite the `$failedIncompatibleApps` if 90 throws an AppManagerException exception. Incompatible apps might actually have been updated. If we move the initialization to the top this should work correctly. How about initializing `$hasNotUpdatedCompatibleApps`, `$failedMissingApps` and `$failedIncompatibleApps` in 73-75 instead of checking `$isMarketEnabled` here. Then you can introduce a new function that does all the error handling in 120-134 and call it from the catch as well.

VicDeo replied:
@butonic This catch means that marketplace is not available. Code starting from this line is executed both for disabled market app and when market app had any issue with connection to marketplace

butonic replied:
yup, and  $this->getAppsFromMarket() for failedIncompatibleApps and failedMissingApps might have succeeded ... than internet dies and  $this->getAppsFromMarket() for failedCompatibleApps throws the Exception ... this would overwrite the failedIncompatibleApps and failedMissingApps arrays, with the initial value. Thats wrong, isn't it?

butonic reviewed (changes requested):
I think the logic can be improved to prevent the output not reflecting that apps have been updated.

VicDeo committed:
Add a note why this is needed

PVince81 commented:
- [x] TEST: web UI also aborts properly

PVince81 commented:
@VicDeo please fix the bugs and address/clarify @butonic's concerns.

Other than that I think we should merge this. Further polishing (better error messages) to be done separately.

PVince81 commented:
- [ ] BUG: web UI still says "These incompatible apps will be disabled" even though we don't. Need to remove this message.

VicDeo commented:
@PVince81 regaring your testcase https://github.com/owncloud/core/pull/27711#issuecomment-296553105
the app is being pulled from market and becomes compatible. 
But it is going to disabled state just as an app with missing code. :(

VicDeo commented:
@PVince81 regrading output https://github.com/owncloud/core/pull/27711#issuecomment-296549352

I'm limited  with  `info` and `warn` at the moment. And `warn` is not appropriate while `info` is not shown with a default verbocity level:
```
php occ upgrade -v
...
2017-04-24T09:09:26+00:00 Repair info: removing duplicate entries from lucene_status
2017-04-24T09:09:26+00:00 Repair step: Put back missing apps
2017-04-24T09:09:26+00:00 Repair info: Fetching app from market: admin_audit
2017-04-24T09:09:26+00:00 Repair info: Fetching app from market: comments
2017-04-24T09:09:26+00:00 Repair info: Fetching app from market: contacts
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: dav
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: enterprise_key
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: federatedfilesharing
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: federation
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: files
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: files_antivirus
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: files_external
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: files_sharing
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: files_texteditor
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: files_trashbin
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: files_versions
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: gallery
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: market
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: provisioning_api
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: systemtags
2017-04-24T09:09:27+00:00 Repair info: Fetching app from market: testing
2017-04-24T09:09:28+00:00 Repair info: Fetching app from market: updatenotification
2017-04-24T09:09:28+00:00 Repair info: Fetching app from market: user_ldap
2017-04-24T09:09:28+00:00 Repair info: Fetching app from market: user_persona
2017-04-24T09:09:28+00:00 Updating database schema
2017-04-24T09:09:28+00:00 Updated database

```

PVince81 commented:
> the app is being pulled from market and becomes compatible

If the version from the market is the same it shouldn't try and pull it.

VicDeo commented:
@PVince81 it's an market app business actually. what I pass from this point is an appid.

VicDeo commented:
indeed. everything that was (re)downloaded  becomes disabled :crying_cat_face:

PVince81 commented:
@VicDeo could it be https://github.com/owncloud/core/pull/27730 ?

VicDeo commented:
@PVince81 yes! Reenabling in not needed at all for any kind of apps with  `php occ upgrade --no-app-disable`

VicDeo committed:
Consider all apps as failed

VicDeo committed:
Remove enabling workaround

VicDeo committed:
Put compatible apps at the very end

PVince81 commented:
Rebased to get the "don't disable third party app" improvement.

Talked with @VicDeo and he said there are only cosmetic changes left to do.

So I'm going to redo the test plan and hope all goes well now.

Commit:
Remove pre-update message about disabled incompatible apps

PVince81 commented:
@VicDeo I removed the obsolete message here: https://github.com/owncloud/core/pull/27711/commits/e4cee4d652d86356a59566cf26bda20e3e7adcd5

PVince81 commented:
Tests:

#### Marketplace unavailable / offline
- [ ] TEST: upgrade with no changes to do: skipped, success
- [ ] TEST: upgrade with outdated app: abort
- [ ] TEST: upgrade with missing app code: abort
- [ ] TEST: admin manually updates local app code: success
- [ ] TEST: admin disables broken app: success

#### Marketplace available but app does not exist there
- [ ] TEST: upgrade with no changes to do: skipped, success
- [ ] TEST: upgrade with outdated app: abort
- [ ] TEST: upgrade with missing app code: abort
- [ ] TEST: admin manually updates local app code: success
- [ ] TEST: admin disables broken app: success

#### Marketplace available, app exists but no new version
- [ ] TEST: upgrade with no changes to do: skipped, success
- [ ] TEST: upgrade with outdated app: abort
- [ ] TEST: upgrade with missing app code: app is downloaded from market, success
- [ ] TEST: admin manually updates local app code: success
- [ ] TEST: admin disables broken app: success

#### Marketplace available, app exists and new version available
- [ ] TEST: upgrade with no changes to do: skipped, success
- [ ] TEST: upgrade with outdated app: app updated, success
- [ ] TEST: upgrade with missing app code: app downloaded from market, success
- [ ] TEST: upgrade still compatible app, and version is available: app updated, success
- [ ] TEST: admin manually updates local app code: success
- [ ] TEST: admin disables broken app: success

PVince81 commented:
- [ ] BUG: only fail once (I pointed the market URL to localhost)

```
Repair warning: GuzzleHttp\Exception\ConnectException
Repair warning: cURL error 7: Failed to connect to marketplace.int.owncloud.com port 443: Connection refused
Repair warning: GuzzleHttp\Exception\ConnectException
Repair warning: cURL error 7: Failed to connect to marketplace.int.owncloud.com port 443: Connection refused
```

To be raised separately.

PVince81 commented:
- [ ] BUG: ⚠️ occ upgrade for apps should not try updating apps

Basically when only a single app has a new version in info.xml, one needs to run occ upgrade to proceed with the DB upgrade. When this happens we **must not** attempt to update the app codes.

Need to add a check for "is core upgrade needed", if yes then also update app codes
If only apps need DB update, proceed without market.

PVince81 commented:
@VicDeo https://github.com/owncloud/core/pull/27711#issuecomment-296553105 still doesn't work correctly.

It seems it decides to disable the outdated app automatically and moves forward. This is very bad because it could disable an important app. This is the case where the app exists in app store but the platform was updated to a more recent incompatible version.

You need to only accept the market version if it is strictly bigger than the current version from info.xml.
However if the app code is missing, it is fine to take the same version.

PVince81 commented:
- [ ] BUG: Is it loading the same apps list repeatedly ???

```
Repair warning: Client error response [url] localhost/api/v1/platform/10.0.0/apps.json [status code] 404 [reason phrase] Not Found
Repair warning: GuzzleHttp\Exception\ClientException
Repair warning: Client error response [url] localhost/api/v1/platform/10.0.0/apps.json [status code] 404 [reason phrase] Not Found
Repair warning: GuzzleHttp\Exception\ClientException
Repair warning: Client error response [url] localhost/api/v1/platform/10.0.0/apps.json [status code] 404 [reason phrase] Not Found
Repair warning: GuzzleHttp\Exception\ClientException
Repair warning: Client error response [url] localhost/api/v1/platform/10.0.0/apps.json [status code] 404 [reason phrase] Not Found
Repair warning: GuzzleHttp\Exception\ClientException
Repair warning: Client error response [url] localhost/api/v1/platform/10.0.0/apps.json [status code] 404 [reason phrase] Not Found
Repair warning: GuzzleHttp\Exception\ClientException
Repair warning: Client error response [url] localhost/api/v1/platform/10.0.0/apps.json [status code] 404 [reason phrase] Not Found
Repair warning: GuzzleHttp\Exception\ClientException
Repair warning: Client error response [url] localhost/api/v1/platform/10.0.0/apps.json [status code] 404 [reason phrase] Not Found
Repair warning: GuzzleHttp\Exception\ClientException
Repair warning: Client error response [url] localhost/api/v1/platform/10.0.0/apps.json [status code] 404 [reason phrase] Not Found
```

Commit:
Only update market apps when doing a core upgrade

If doing an app-only upgrade then it's not about updating the code but
about updating the database.

Still keep the logic about missing app code and incompatible apps.

PVince81 commented:
@VicDeo ok you were right. The fact that the old app got reinstalled even if it had the same version was a side-effect of my testing. Now when testing this case I increase the core version to "10.1" and then it refetches the market app and sees that it's incompatible.

The sad part is that it refetches the app needlessly, but we can live with this for now and optimize this separately.

PVince81 commented:
- [ ] BUG: something is **still** disabling the "contacts" app after an upgrade and even after the code has been replaced.

It is likely that the app versions are cached somewhere and used by the old code that disabled incompatible apps.

PVince81 commented:
- [ ] To raise: add OCC command "market:upgrade" to upgrade all apps like we did here
- [ ] To raise: argument to force disabling apps `occ upgrade --disable-incompatible` instead of aborting

Commit:
Clear $appInfo cache after updating app

PVince81 commented:
Fixed issue with incompatible app getting autodisabled even after updating to a compatible one: https://github.com/owncloud/core/pull/27711/commits/f3607c006ea13a1f52ef4351fcb44181d3119c0d
(hint: caching...)

PVince81 commented:
I fixed the repeated Guzzle issues by properly catching `ClientException` here: https://github.com/owncloud/market/pull/33/commits/ad482cf250ea49a56568d17786d1eef19ffe921f

PVince81 commented:
@VicDeo @butonic can you guys check my latest code changes ? Thanks

Commit:
Output reason when marketplace not accessible during upgrade

VicDeo committed:
Catch AppUpdateNotFoundException

VicDeo commented:
@PVince81 Additionally fixed: update to the same version. 
It was even capable to downgrade the app so I added an additional check into the listener.
